src/scram.c

```diff
@@ -841,33 +841,48 @@ static bool build_mock_scram_secret(const char *username, ScramState *scram_stat
 	return false;
 }
 
-char *build_server_first_message(ScramState *scram_state, const char *username, const char *stored_secret)
+char *build_server_first_message(ScramState *scram_state, PgCredentials *user, const char *stored_secret)
 {
 	uint8_t raw_nonce[SCRAM_RAW_NONCE_LEN + 1];
 	int encoded_len;
 	size_t len;
 	char *result;
 
 	if (!stored_secret) {
-		if (!build_mock_scram_secret(username, scram_state))
+		if (!build_mock_scram_secret(user->name, scram_state))
 			goto failed;
 	} else {
-		switch (get_password_type(stored_secret)) {
-		case PASSWORD_TYPE_SCRAM_SHA_256:
-			if (!parse_scram_secret(stored_secret,
-						&scram_state->iterations,
-						&scram_state->salt,
-						scram_state->StoredKey,
-						scram_state->ServerKey))
+		if (user->adhoc_scram_secrets_cached) {
+			scram_state->iterations = user->scram_Iiterations;
+			scram_state->salt = strdup(user->scram_SaltKey);
+			memcpy(scram_state->StoredKey, user->scram_StoredKey, sizeof(user->scram_StoredKey));
+			memcpy(scram_state->ServerKey, user->scram_ServerKey, sizeof(user->scram_ServerKey));
+		} else {
+			switch (get_password_type(stored_secret)) {
+			case PASSWORD_TYPE_SCRAM_SHA_256:
+				if (!parse_scram_secret(stored_secret,
+							&scram_state->iterations,
+							&scram_state->salt,
+							scram_state->StoredKey,
+							scram_state->ServerKey))
+					goto failed;
+				break;
+			case PASSWORD_TYPE_PLAINTEXT:
+				if (!build_adhoc_scram_secret(stored_secret, scram_state))
+					goto failed;
+				break;
+			default:
+				/* shouldn't get here */
 				goto failed;
-			break;
-		case PASSWORD_TYPE_PLAINTEXT:
-			if (!build_adhoc_scram_secret(stored_secret, scram_state))
-				goto failed;
-			break;
-		default:
-			/* shouldn't get here */
-			goto failed;
+			}
+
+			if (!user->dynamic_passwd) {
```

Comment:
I think there's probably some possibility of even caching this for dynamic_passwd users (aka `auth_user`), but that can be done in a follow up PR.